### PR TITLE
feat(generate)!: initial generate ssp command

### LINKED ISSUE #803: Generate SSP Command
### Is your feature request related to a problem? Please describe.

Add the `lula generate ssp <flags>` command

### Describe the solution you'd like

- **Given** component definition(s)
- **When** lula generate ssp -f component-definition(s) <flags>
- **Then** SSP is generated

### Additional context

Blocked by #797 

### Expected Deliverable

- [ ] Cmd + e2e cmd tests
- [ ] CLI Docs updated

### PR DESCRIPTION
**Pull request by meganwolf0**
## Description

Initial generate ssp command
* Adds `lula generate system-security-plan` with required profile and optional component path(s)
* (**Breaking**) Modifies the library ssp functionality to incorporate profile control resolution, instead of depending on controls defined by the component definitions (more information in below doc)
* Updates the `docs/oscal/system-security-plan.md` to reflect logic followed in the generate ssp command

Follow-on issues include:
- #810 - "Partial" compose is most likely the disaggregation of importing components to the component def and imported validations to the back-matter (follow-on discussion is potentially around continued import of validations to back-matter and utility)
- #624 - Injecting external, user defined, fields into the SSP (or other generated OSCAL models)

## Related Issue

Fixes #803

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [X] Test, docs, adr added or updated as needed
- [ ] (N/A) [Schema Updates](https://github.com/defenseunicorns/lula/blob/main/docs/community-and-contribution/schema-updates.md) applied
- [X] [Contributor Guide Steps](https://github.com/defenseunicorns/lula/blob/main/CONTRIBUTING.md) followed

### CODE REVIEW

File: docs/cli-commands/lula_generate_system-security-plan.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+description: Lula CLI command reference for <code>lula generate system-security-plan</code>.
+type: docs
+---
+## lula generate system-security-plan
+
+Generate a system security plan OSCAL template
+
+### Synopsis
+
+Generation of a System Security Plan OSCAL artifact from a source profile along with an optional list of component definitions.
```
Comment (mildwonkey) [inline docs/cli-commands/lula_generate_system-security-plan.md:12]: this is so nitpicky i'm sorry, but is it a template, or an artifact? Should this use the same terminology as the sentence above?

File: docs/oscal/system-security-plan.md [hunk 2]
```diff
@@ -1,25 +1,33 @@
 # System Security Plan
```
Comment (mildwonkey) [inline docs/oscal/system-security-plan.md:2]: this is a really solid doc overall 👍🏻

File: src/test/e2e/cmd/generate_ssp_test.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+		// Create the new ssp object
+		err = ssp.NewModel(compiledBytes)
+		require.NoError(t, err, "error creating oscal model from ssp artifact")
+
+		complete := ssp.GetCompleteModel()
+		sspModel := complete.SystemSecurityPlan
+
+		require.NotNil(t, sspModel, "expected the SystemSecurityPlan model to be non-nil")
+		assert.Equal(t, 3, len(sspModel.ControlImplementation.ImplementedRequirements), "expected 3 controls")
```
Comment (mildwonkey) [inline src/test/e2e/cmd/generate_ssp_test.go:131]: I have no objection and certainly don't want anything you've written removed, but fwiw adding error messages to the asserts is rarely valuable IMO - if our error messages aren't clear enough to us you why the test failed, that's a sign to fix the error message in the test

**Review by mildwonkey — COMMENTED** [review]
I don't necessarily feel qualified to approve this (the oscal concepts are still so new to me), but I also don't have any objections and it all looks reasonable to me :) I left some comments but nothing that requires changes; feel free to dismiss 'em ❤️

File: docs/oscal/system-security-plan.md [hunk 4]
```diff
@@ -1,25 +1,33 @@
...
+A [System Security Plan](https://pages.nist.gov/OSCAL/resources/concepts/layer/implementation/ssp/) is an OSCAL-specific model to represent a system as a whole. In Lula, the `generate system-security-plan` command creates a representative `system-security-plan` object to describe the system as a whole. The only required input to the command is an OSCAL `profile` that defines the controls the system will implement. Optionally, one to many `component-definitions` can be provided to include the details of the components that make up the system and the respective implementation of the controls.
```
Comment (brandtkeller) [inline docs/oscal/system-security-plan.md]: "system as a whole" feels like potentially repetitive language. We can probably drop that in favor of "the system".

File: src/cmd/generate/system-security-plan.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+To specify the name and filetype of the generated artifact:
+	lula generate system-security-plan -p <path/to/profile> -c <path/to/component-definition> -o my_ssp.yaml
+`
+
+var sspLong = `Generation of a System Security Plan OSCAL artifact from a source profile along with an optional list of component definitions.`
+
+func GenerateSSPCommand() *cobra.Command {
+	var (
+		component  []string
```
Comment (brandtkeller) [inline src/cmd/generate/system-security-plan.go]: [nit] Would plural `components` be more intuitive to the user?

**Review by brandtkeller — COMMENTED** [review]
Few minor comments thus far - have a little more tracing to do - otherwise looking great.

**Comment by brandtkeller** [comment]
(few conflicts to resolve when you have a minute)

File: docs/oscal/system-security-plan.md (hunk 2 above)
Comment (meganwolf0) [reply docs/oscal/system-security-plan.md:2]: Thank you! @CloudBeard created this doc so props to him! I just updated some of it to reflect the new generation assumption 😄

File: src/cmd/generate/system-security-plan.go (hunk 5 above)
Comment (meganwolf0) [reply src/cmd/generate/system-security-plan.go]: fair nit, I think you're right - will update

File: docs/oscal/system-security-plan.md (hunk 4 above)
Comment (meganwolf0) [reply docs/oscal/system-security-plan.md]: updating!

File: docs/cli-commands/lula_generate_system-security-plan.md (hunk 1 above)
Comment (meganwolf0) [reply docs/cli-commands/lula_generate_system-security-plan.md:12]: This is a fair nitpicky comment - and I think I'd prefer the language `artifact` as it's a valid OSCAL SSP. Although I do see in generate component and profile we used the language of `template` (which is where I copy-pasta'd from)... Template now has a very specific denotation within Lula so I'm not really sure that's the right language to use and we should probably consolidate around another phrase for the `generate` commands. 

@brandtkeller might have some thoughts.

File: src/test/e2e/cmd/generate_ssp_test.go (hunk 3 above)
Comment (meganwolf0) [reply src/test/e2e/cmd/generate_ssp_test.go:131]: That's fair - I generally haven't been adding error messages and tbh I'm not really sure what came over me on this one that provided the inclination to adding them 😅 
That is a good point about these err messages potentially obfuscating _bad_ error output from the functions though 🤔

**Comment by meganwolf0** [comment]
**Update**
* Resolved conflicts on main (with oscalTypes)
* Fixed some doc words
* Modified flag/var name for `component` -> `components` in generate SSP
* `template` -> `artifact` in short help for both SSP and Profile (Do we want to update generate component text as well OR is there another word we should give these things?)

Note - the `help.golden` files are kind of annoying since we have to re-generate the cli-docs + re-run the tests to generate the help docs when some command text words change... just throwing out the question of if those golden checks on `help` are valuable?

File: docs/cli-commands/lula_generate_system-security-plan.md (hunk 1 above)
Comment (brandtkeller) [reply docs/cli-commands/lula_generate_system-security-plan.md:12]: I would agree with your assessment. `artifact` is more aligned to the story Lula should tell in the long run. We've described it as a template in the past - as it's not a "complete" artifact - but it is still an artifact nonetheless.

File: src/cmd/generate/system-security-plan.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+					return fmt.Errorf("error creating new composer: %v", err)
+				}
+
+				model, err := composer.ComposeFromPath(cmd.Context(), componentPath)
+				if err != nil {
+					return fmt.Errorf("error composing model from path: %v", err)
+				}
+
+				componentDefs = append(componentDefs, model.ComponentDefinition)
+				command += fmt.Sprintf(" --component %s", componentPath)
```
Comment (brandtkeller) [inline src/cmd/generate/system-security-plan.go]:
```suggestion
				command += fmt.Sprintf(" --components %s", componentPath)
```

**Review by brandtkeller — CHANGES_REQUESTED** [review]
besides a typo I feel like this looks great and really sets us up well for AP generation.

File: src/cmd/generate/system-security-plan.go (hunk 6 above)
Comment (meganwolf0) [reply src/cmd/generate/system-security-plan.go]: ahh a `component` hold out, nice catch!

**Review by mildwonkey — APPROVED** [review]
🎉

**Review by brandtkeller — APPROVED** [review]

**Pull request merged**